src/openms/source/ANALYSIS/ID/AhoCorasickDA.cpp

```diff
@@ -0,0 +1,820 @@
+    }
+
+
+    UInt32 num_emps_ = 0;
+
+    if (empty_head_ != 0) {
+      auto pos = empty_head_ ;
+      do
+      {
+        pos = build_info_[pos].next;
+        ++num_emps_;
+      } while (empty_head_  != pos);
+    }
+
+    UInt32 sub_str = 0;
+    for (auto l : bc_)
+    {
+      if (l.leaf_flag)
+      {
+        sub_str += l.term_flag;
+      }
+
+    }
+    std::cout << "num_emps_ " << num_emps_ << std::endl;
+    std::cout << "sub_str " << sub_str << std::endl;
+    std::cout << "# Strings: " << sequences_.size() << "  -  Found: " << count << " Out: " << out << std::endl;
+
+  }
+
+  bool AhoCorasickDA::retrievalCDA()
+  {
+    while (!bc_[node_pos_].leaf_flag)
+    {
+      if (*protein_ == '\0')
+      {
+        return bc_[node_pos_].term_flag == 1;
+      }
+      node_pos_ = child_(bc_[node_pos_].base, *protein_);
+      if(bc_[node_pos_].lcheck != *protein_)
+      {
+        return false;
+      }
+      protein_++;
+    }
+    if (*protein_ == '\0')
+    {
+      protein_++;
+      return bc_[node_pos_].term_flag == 1;
+    }
+    auto tail_str = &tail_[bc_[node_pos_].base];
+    node_pos_ = -bc_[node_pos_].base;
+    while (*tail_str != '\0' && *tail_str == *protein_)
+    {
+      ++tail_str;
+      ++protein_;
+      --node_pos_;
+    }
+    ++node_pos_;
+    return *tail_str == *protein_;
+  }
+
+
+
+
+
+  //-------------------------------------------------------------------------------------------------------------------------------------
+  //  CDA
+  //-------------------------------------------------------------------------------------------------------------------------------------
+
+
+
+
+  UInt32 AhoCorasickDA::child_(const UInt32& base_value, const char& label)
+  {
+    return base_value ^ code_(label);
+    //return base_value + code_(label);
+  }
+
+  UInt32 AhoCorasickDA::code_(const char& label)
+  {
+    return code_tabel_[static_cast<UInt>(label)];
+  }
+
+
+  void AhoCorasickDA::setProtein(const char* prot)
+  {
+    protein_ = prot;
+    prot_pos_ = 0;
+    node_pos_ = 0;
+  }
+
+  bool AhoCorasickDA::testAmbiguous(const char aa)
+  {
+    if (!isAmbiguous_(aa))
+    {
+      return true;
+    }
+
+    Spawn cur_spawn;
+    
+    // B = D, N
+    if (code_(aa) == 22)
+    {
+      cur_spawn.a_acids.push_back(code_('D'));
+      cur_spawn.a_acids.push_back(code_('N'));
+    }
+
+    // J = I, L
+    if (code_(aa) == 24)
+    {
+      cur_spawn.a_acids.push_back(code_('I'));
+      cur_spawn.a_acids.push_back(code_('L'));
+    }
+
+    // Z = E, Q
+    if (code_(aa) == 23)
+    {
+      cur_spawn.a_acids.push_back(code_('E'));
+      cur_spawn.a_acids.push_back(code_('Q'));
+    }
+
+    if (code_(aa) == 25)
+    {
+      cur_spawn.a_acids.resize(22);
+      std::iota(cur_spawn.a_acids.begin(), cur_spawn.a_acids.end(), 0);
+    }
+
+    while (!cur_spawn.a_acids.empty())
+    {
+      Int32 child_pos = getNode_(node_pos_, cur_spawn.a_acids[0]);
+      cur_spawn.a_acids.erase(cur_spawn.a_acids.begin());
```

Comment:
inefficient.
just iterate over the possible AA's

Reply:
This code is not possible anyway. Just writing down my first thought about ambiguous aa.